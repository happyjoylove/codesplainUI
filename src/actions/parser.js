export const LOAD_PARSER = 'LOAD_PARSER';
export const ADD_ERROR = 'ADD_ERROR';
export const CLEAR_ERRORS = 'CLEAR_ERRORS';

export const loadParser = language => ({
  type: LOAD_PARSER,
  meta: {
    WebWorker: true,
  },
  payload: language,
});

export const addError = error => ({
  type: ADD_ERROR,
  payload: error,
});

export const clearErrors = () => ({
  type: CLEAR_ERRORS,
});
