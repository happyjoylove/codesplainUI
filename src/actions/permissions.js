export const SET_PERMISSIONS = 'SET_PERMISSIONS';
export const SET_AUTHOR = 'SET_AUTHOR';

export const setPermissions = permissions => ({
  type: SET_PERMISSIONS,
  payload: permissions,
});

export const setAuthor = author => ({
  type: SET_AUTHOR,
  payload: author,
});
