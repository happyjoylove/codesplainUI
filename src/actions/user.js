import axios from 'axios';
import mapValues from 'lodash/mapValues';

import { makeSaveEndpointUrl, sanitizeSnippetList } from '../util/requests';

const API_URL = process.env.REACT_APP_API_URL;

// Util func to check for 'NoSuchKey' responses from S3
const noSuchKey = data => (
  typeof data === 'string' && data.includes('NoSuchKey')
);

export const ADD_ORG = 'ADD_ORG';
export const ADD_ORGANIZATIONS = 'ADD_ORGANIZATIONS';
export const CLEAR_USER_CREDENTIALS = 'CLEAR_USER_CREDENTIALS';
export const RESTORE_USER_CREDENTIALS = 'RESTORE_USER_CREDENTIALS';
export const SAVE_ACCESS_TOKEN = 'SAVE_ACCESS_TOKEN';
export const SAVE_USERNAME = 'SAVE_USERNAME';
export const SET_AVATAR_URL = 'SET_AVATAR_URL';
export const SET_SNIPPET_LISTS = 'SET_SNIPPET_LISTS';
export const SET_USER_SNIPPETS = 'SET_USER_SNIPPETS';
export const SWITCH_ORG = 'SWITCH_ORG';
export const UPDATE_USER_SNIPPETS = 'UPDATE_USER_SNIPPETS';

export const setAvatarUrl = url => ({
  type: SET_AVATAR_URL,
  payload: url,
});

export const addOrg = org => ({
  type: ADD_ORG,
  payload: org,
});

export const addOrganizations = organizations => ({
  type: ADD_ORGANIZATIONS,
  payload: organizations,
});

export const clearUserCredentials = () => ({
  type: CLEAR_USER_CREDENTIALS,
});

export const saveAccessToken = accessToken => ({
  type: SAVE_ACCESS_TOKEN,
  payload: accessToken,
});

export const saveUsername = username => ({
  type: SAVE_USERNAME,
  payload: username,
});

export const setSnippetLists = snippetLists => ({
  type: SET_SNIPPET_LISTS,
  payload: snippetLists,
});

export const setUserSnippets = snippetMeta => ({
  type: SET_USER_SNIPPETS,
  payload: snippetMeta,
});

export const switchOrg = org => ({
  type: SWITCH_ORG,
  payload: org,
});

export const updateUserSnippets = () => (dispatch, getState) => {
  const {
    user: {
      token,
      username,
    },
  } = getState();
  // Fetch the user's snippet meta data and save it
  const headers = {
    Accept: 'application/json',
    Authorization: `token ${token}`,
  };
  return axios.get(makeSaveEndpointUrl(username), { headers })
    .then(({ data }) => {
      // Jump to catch block if the user has no index.json file:
      if (noSuchKey(data)) {
        throw new Error(`index.json does not exist for ${username}!`);
      }
      dispatch(setUserSnippets(data));
    });
};

export const fetchSnippetLists = () => (dispatch, getState) => {
  const {
    token,
    orgs,
  } = getState().user;
  const headers = {
    Authorization: `token ${token}`,
  };
  const url = `${API_URL}/users?users=${orgs.join(',')}`;
  // Remove the username from the org list
  return axios({
    method: 'GET',
    url,
    headers,
  })
    .then(({ data: snippetLists }) => {
      const sanitizedSnippetLists = mapValues(snippetLists, sanitizeSnippetList);
      dispatch(setSnippetLists(sanitizedSnippetLists));
    });
};

export const fetchAccessToken = authCode => () => {
  const reqUrl = `${API_URL}/auth`;
  const reqBody = { code: authCode };
  return axios({
    method: 'POST',
    url: reqUrl,
    data: reqBody,
  });
};

export const fetchUserInfo = () => (dispatch, getState) => {
  const { token } = getState().user;
  const reqHeaders = {
    Accept: 'application/json',
    Authorization: `token ${token}`,
  };
  return axios({
    method: 'GET',
    url: 'https://api.github.com/user',
    headers: reqHeaders,
  });
};

export const fetchUserOrgs = () => (dispatch, getState) => {
  const { token } = getState().user;
  const reqHeaders = {
    Accept: 'application/json',
    Authorization: `token ${token}`,
  };
  return axios({
    method: 'GET',
    url: 'https://api.github.com/user/orgs',
    headers: reqHeaders,
  });
};
