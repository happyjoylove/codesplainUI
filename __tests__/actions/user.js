import moxios from 'moxios';

import * as actions from '../../src/actions/user';
import generateMockStore from '../../src/testUtils/mockStore';

describe('Actions: User', () => {
  describe('action creators', () => {
    it('creates correct ADD_ORG object', () => {
      const org = 'galactic-federation';
      expect(actions.addOrg(org)).toMatchSnapshot();
    });

    it('creates correct ADD_ORGANIZATIONS object', () => {
      const organizations = ['Galactic Federation', 'Council Of Ricks'];
      expect(actions.addOrganizations(organizations)).toMatchSnapshot();
    });

    it('creates correct CLEAR_USER_CREDENTIALS object', () => {
      expect(actions.clearUserCredentials()).toMatchSnapshot();
    });

    it('creates correct SET_AVATAR_URL object', () => {
      const url = 'https://foobar.com/quxbaz';
      expect(actions.setAvatarUrl(url)).toMatchSnapshot();
    });

    it('creates correct SAVE_USERNAME object', () => {
      const token = 'token';
      expect(actions.saveUsername(token)).toMatchSnapshot();
    });

    it('creates correct SAVE_ACCESS_TOKEN object', () => {
      const token = 'token';
      expect(actions.saveAccessToken(token)).toMatchSnapshot();
    });

    it('creates correct SAVE_USERNAME object', () => {
      const token = 'token';
      expect(actions.saveUsername(token)).toMatchSnapshot();
    });

    it('creates correct SET_SNIPPET_LISTS object', () => {
      const snippetLists = ['list', 'of', 'snippets'];
      expect(actions.setSnippetLists(snippetLists)).toMatchSnapshot();
    });

    it('creates correct SET_USER_SNIPPETS object', () => {
      const snippetLists = ['list', 'of', 'snippets'];
      expect(actions.setUserSnippets(snippetLists)).toMatchSnapshot();
    });

    it('creates correct SWITCH_ORG object', () => {
      const org = 'galactic-federation';
      expect(actions.switchOrg(org)).toMatchSnapshot();
    });
    it('creates correct SET_USER_SNIPPETS object', () => {
      const snippetsList = [
        {
          snippetName: 'Test Snippet',
          language: 'Python 3',
          lastEdited: '2017-04-05T12:52:20.099Z',
          private: true,
        },
      ];
      expect(actions.setUserSnippets(snippetsList)).toMatchSnapshot();
    });
  });

  describe('async actions', () => {
    beforeEach(() => {
      moxios.install();
    });

    afterEach(() => {
      moxios.uninstall();
    });

    describe('UPDATE_USER_SNIPPETS', () => {
      it('dispatches correct actions if successful', () => {
        moxios.wait(() => {
          const request = moxios.requests.mostRecent();
          request.respondWith({
            status: 200,
            response: {
              test_snippet_1: {
                snippetName: 'Test Snippet 1',
                language: 'Python 3',
                lastEdited: '2017-04-05T12:52:20.099Z',
                private: true,
              },
              test_snippet_2: {
                snippetName: 'Test Snippet 2',
                language: 'Python 3',
                lastEdited: '2017-04-05T12:52:20.099Z',
                private: true,
              },
            },
          });
        });
        const store = generateMockStore({
          user: {
            username: 'FooBar',
            token: '1234',
          },
        });
        return store.dispatch(actions.updateUserSnippets())
          .then(() => { // return of async actions
            expect(store.getActions()).toMatchSnapshot();
          });
      });

      it('dispatches UPDATE_USER_SNIPPETS_FAILED if unsuccessful', () => {
        moxios.wait(() => {
          const request = moxios.requests.mostRecent();
          request.respondWith({
            status: 400,
          });
        });
        const store = generateMockStore({
          user: {
            username: 'FooBar',
            token: '1234',
          },
        });
        return store.dispatch(actions.updateUserSnippets())
          .catch(() => {
            expect(store.getActions()).toMatchSnapshot();
          });
      });
    });
  });
});
