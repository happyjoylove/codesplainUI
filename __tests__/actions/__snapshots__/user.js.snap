exports[`Actions: User action creators creates correct ADD_ORG object 1`] = `
Object {
  "payload": "galactic-federation",
  "type": "ADD_ORG",
}
`;

exports[`Actions: User action creators creates correct ADD_ORGANIZATIONS object 1`] = `
Object {
  "payload": Array [
    "Galactic Federation",
    "Council Of Ricks",
  ],
  "type": "ADD_ORGANIZATIONS",
}
`;

exports[`Actions: User action creators creates correct CLEAR_USER_CREDENTIALS object 1`] = `
Object {
  "type": "CLEAR_USER_CREDENTIALS",
}
`;

exports[`Actions: User action creators creates correct SAVE_ACCESS_TOKEN object 1`] = `
Object {
  "payload": "token",
  "type": "SAVE_ACCESS_TOKEN",
}
`;

exports[`Actions: User action creators creates correct SAVE_USERNAME object 1`] = `
Object {
  "payload": "token",
  "type": "SAVE_USERNAME",
}
`;

exports[`Actions: User action creators creates correct SAVE_USERNAME object 2`] = `
Object {
  "payload": "token",
  "type": "SAVE_USERNAME",
}
`;

exports[`Actions: User action creators creates correct SET_AVATAR_URL object 1`] = `
Object {
  "payload": "https://foobar.com/quxbaz",
  "type": "SET_AVATAR_URL",
}
`;

exports[`Actions: User action creators creates correct SET_SNIPPET_LISTS object 1`] = `
Object {
  "payload": Array [
    "list",
    "of",
    "snippets",
  ],
  "type": "SET_SNIPPET_LISTS",
}
`;

exports[`Actions: User action creators creates correct SET_USER_SNIPPETS object 1`] = `
Object {
  "payload": Array [
    "list",
    "of",
    "snippets",
  ],
  "type": "SET_USER_SNIPPETS",
}
`;

exports[`Actions: User action creators creates correct SET_USER_SNIPPETS object 2`] = `
Object {
  "payload": Array [
    Object {
      "language": "Python 3",
      "lastEdited": "2017-04-05T12:52:20.099Z",
      "private": true,
      "snippetName": "Test Snippet",
    },
  ],
  "type": "SET_USER_SNIPPETS",
}
`;

exports[`Actions: User action creators creates correct SWITCH_ORG object 1`] = `
Object {
  "payload": "galactic-federation",
  "type": "SWITCH_ORG",
}
`;

exports[`Actions: User async actions UPDATE_USER_SNIPPETS dispatches UPDATE_USER_SNIPPETS_FAILED if unsuccessful 1`] = `Array []`;

exports[`Actions: User async actions UPDATE_USER_SNIPPETS dispatches correct actions if successful 1`] = `
Array [
  Object {
    "payload": Object {
      "test_snippet_1": Object {
        "language": "Python 3",
        "lastEdited": "2017-04-05T12:52:20.099Z",
        "private": true,
        "snippetName": "Test Snippet 1",
      },
      "test_snippet_2": Object {
        "language": "Python 3",
        "lastEdited": "2017-04-05T12:52:20.099Z",
        "private": true,
        "snippetName": "Test Snippet 2",
      },
    },
    "type": "SET_USER_SNIPPETS",
  },
]
`;
